params/upgrades_config.go

```diff
@@ -0,0 +1,252 @@
+
+		for _, getter := range getters {
+			if config := getter(&upgrade); isNil(config) {
+				continue
+			}
+			if hasKey {
+				return fmt.Errorf("PrecompileUpgrades[%d] has more than one key set", i)
+			}
+			hasKey = true
+		}
+	}
+
+	for _, getter := range getters {
+		var (
+			lastUpgraded *big.Int
+			disabled     bool
+		)
+		// check the genesis chain config for any enabled upgrade
+		if config := getter(&c.Upgrade); !isNil(config) {
+			disabled = false
+			lastUpgraded = config.Timestamp()
+		} else {
+			disabled = true
+		}
+		// next range over UpgradesFromUpgradeBytes
+		// to verify correct use of disabled and blockTimestamps.
+		for i, upgrade := range c.PrecompileUpgrades {
+			if config := getter(&upgrade); !isNil(config) {
+				if disabled == config.IsDisabled() {
+					return fmt.Errorf("PrecompileUpgrades[%d] disable should be [%v]", i, !disabled)
+				}
+				if lastUpgraded != nil && (config.Timestamp() == nil || lastUpgraded.Cmp(config.Timestamp()) > 0) {
+					return fmt.Errorf("PrecompileUpgrades[%d] timestamp should not be less than [%v]", i, lastUpgraded)
+				}
+
+				disabled = config.IsDisabled()
+				lastUpgraded = config.Timestamp()
+			}
+		}
+	}
+
+	return nil // successfully verified all conditions
+}
+
+// getActiveConfig returns the most recent config that has
+// already forked, or nil if none have been configured or
+// if none have forked yet.
+func (c *UpgradesConfig) getActiveConfig(blockTimestamp *big.Int, getter getterFn) precompile.StatefulPrecompileConfig {
+	configs := c.getActiveConfigs(nil, blockTimestamp, getter)
+	if len(configs) == 0 {
+		return nil
+	}
+	return configs[len(configs)-1] // return the most recent config
+}
+
+// getActiveConfigs returns all forks configured to activate in the
+// specified [from, to] range, in the order of activation.
+func (c *UpgradesConfig) getActiveConfigs(from *big.Int, to *big.Int, getter getterFn) []precompile.StatefulPrecompileConfig {
+	configs := make([]precompile.StatefulPrecompileConfig, 0)
+	// first check the embedded [upgrade] for precompiles configured
+	// in the genesis chain config.
+	if config := getter(&c.Upgrade); !isNil(config) {
+		if utils.IsForkTransition(config.Timestamp(), from, to) {
+			configs = append(configs, config)
+		}
+	}
+	// loop on all upgrades
+	for _, upgrade := range c.PrecompileUpgrades {
+		if config := getter(&upgrade); !isNil(config) {
+			// check if fork is activating in the specified range
+			if utils.IsForkTransition(config.Timestamp(), from, to) {
+				configs = append(configs, config)
+			}
+		}
+	}
+	return configs
+}
+
+// GetContractDeployerAllowListConfig returns the latest forked ContractDeployerAllowListConfig
+// specified by [c] or nil if it was never enabled.
+func (c *UpgradesConfig) GetContractDeployerAllowListConfig(blockTimestamp *big.Int) *precompile.ContractDeployerAllowListConfig {
+	getter := func(u *Upgrade) precompile.StatefulPrecompileConfig { return u.ContractDeployerAllowListConfig }
```

Comment:
The same function is defined above. Can we make these variables above and re-use them in these getter functions?